### PR TITLE
Add support for Service/Folder options.

### LINKED ISSUE #122: Add modes for the folders
According to the documentation, we should be able to set modes (read only, read and write or write only) for a folder, at the service level.

### PR DESCRIPTION
**Pull request by wapiflapi**
This adds support for options at the service/folder level in the configuration files.
Also adds the mode option and fixes #122.

### CODE REVIEW

File: onitu/referee/folder.py [hunk 1]
```diff
@@ -31,82 +18,115 @@ def __repr__(self):
...
-            self.logger.info(
-                "Ignoring event for '{}' in folder {} due to its mimetype: {}",
-                filename, self.name, metadata['mimetype']
-            )
-            return
+        # TODO: We might want to consider doing different checks for the
+        # source than what we do for the targets because in some cases
+        # we might want to cheat. Eg: what happens if the size grows to
+        # large for the source's service, we'dd still like for it to be
+        # copied. Or not? Maybe not. I don't know. Not doing it now.
```
Comment (k4nar) [inline onitu/referee/folder.py]: Can you remove this comment ? :)

File: onitu/referee/folder.py [hunk 2]
```diff
@@ -31,82 +18,115 @@ def __repr__(self):
...
+        def check_list(l, c):
+            if l is None:
+                return None
+            return any(fnmatchcase(c, rule) for rule in l)
 
-        if self.max_size is not None and size > self.max_size:
+        # mode
+
+        modeopts = options.get("mode", "rw")
+        if any(c not in modeopts for c in mode):
```
Comment (k4nar) [inline onitu/referee/folder.py:88]: I'm not sure to understand what this actually does.

File: onitu/referee/folder.py [hunk 3]
```diff
@@ -31,82 +18,115 @@ def __repr__(self):
...
             return False
 
-        return True
+        # min/max size
+
+        minsz = options.get("file_size", {}).get("min")
+        maxsz = options.get("file_size", {}).get("max")
+
+        if minsz is not None and metadata['size'] < self._to_bytes(minsz):
+            self.logger.info(
```
Comment (k4nar) [inline onitu/referee/folder.py:100]: Those messages are not always valid, as you could be ignoring it just for a service.

File: onitu/referee/folder.py [hunk 4]
```diff
@@ -31,82 +18,115 @@ def __repr__(self):
...
+
+        return targets
+
+    def assert_options(self, options, metadata, mode=""):
+
+        # Those options are either our own or those of a service.
+        # This is because there is no difference between
+        # service/folder and folder options.
+
+        def check_list(l, c):
```
Comment (k4nar) [inline onitu/referee/folder.py:70]: I think this function would be better of as a method, with an explicit name.

File: onitu/referee/folder.py (hunk 2 above)
Comment (wapiflapi) [inline onitu/referee/folder.py:88]: `mode` is an argument to the `assert_options` method telling it if we are checking in order to read `r` or write `w` the file. These particular lines check if any character in mode isn't in the allowed mode specified in the options.
This way we are consistent regarding the syntax for modes everywhere and one could pass `rw` as argument to the `assert_options` method without problems if this is needed some time.

File: tests/unit/test_referee_folder.py [hunk 5]
```diff
@@ -1,204 +1,253 @@
 from onitu.referee.folder import Folder
 
 
-def check_size(size, length):
-        folder = Folder('folder', (), None,
-                        file_size={'min': size, 'max': size})
-        assert folder.min_size == length
-        assert folder.max_size == length
+class FakeLogger(object):
```
Comment (wapiflapi) [inline tests/unit/test_referee_folder.py]: I didn't mean to include this in the PR. Forgot to fix it.
The problem is a Folder takes a logger and it is needed. How do we give it a Logger from within a unit test? Is there a best practice for this?

Comment (k4nar) [inline tests/unit/test_referee_folder.py]: Nope, unit tests are very confidential right now, so we don't really have any tools for them. I guess you could still use a regular logger from Logbook.

**Pull request merged**